src/vumi2/applications/junebug_message_api/junebug_message_api.py

```diff
@@ -0,0 +1,163 @@
+from logging import getLogger
+from typing import Optional
+
+from attrs import define, field
+from httpx import AsyncClient
+
+from vumi2.cli import class_from_string
+from vumi2.messages import DeliveryStatus, Event, EventType, Message
+from vumi2.workers import BaseConfig, BaseWorker
+
+from . import junebug_state_cache
+
+LOG_MSG_HTTP_ERR = (
+    "Error sending message, received HTTP code %(code)s"
+    " with body %(body)s. Message: %(message)s"
+)
+LOG_EV_HTTP_ERR = (
+    "Error sending event, received HTTP code %(code)s"
+    " with body %(body)s. Event: %(event)s"
+)
+LOG_EV_URL_MISSING = "Cannot find event URL, missing user_message_id: %(event)s"
+
+logger = getLogger(__name__)
+
+
+def junebug_from_msg(message: Message) -> dict:
+    vumi_dict = message.serialise()
+    msg = {
+        "to": vumi_dict["to_addr"],
+        "from": vumi_dict["from_addr"],
+        "group": vumi_dict["group"],
+        "message_id": vumi_dict["message_id"],
+        "channel_id": vumi_dict["transport_name"],
+        "timestamp": vumi_dict["timestamp"],
+        "reply_to": vumi_dict["in_reply_to"],
+        "content": vumi_dict["content"],
+        "channel_data": vumi_dict["helper_metadata"],
+    }
+    msg["channel_data"]["session_event"] = vumi_dict["session_event"]
+    return msg
+
+
+def junebug_from_ev(event: Event, channel_id: str) -> dict:
+    vumi_dict = event.serialise()
+    ev = {
+        "channel_id": channel_id,
+        "timestamp": vumi_dict["timestamp"],
+        "event_details": {},
+        "event_type": None,
+    }
+    if event.event_type == EventType.ACK:
+        ev["event_type"] = "submitted"
+    elif event.event_type == EventType.NACK:
+        ev["event_type"] = "rejected"
+        ev["event_details"] = {"reason": event.nack_reason}
+    elif event.event_type == EventType.DELIVERY_REPORT:
+        ev["event_type"] = {
+            DeliveryStatus.PENDING: "delivery_pending",
+            DeliveryStatus.FAILED: "delivery_failed",
+            DeliveryStatus.DELIVERED: "delivery_succeeded",
+            None: None,
+        }[event.delivery_status]
+    return ev
+
+
+@define(kw_only=True)
+class JunebugMessageApiConfig(BaseConfig):
+    connector_name: str
+
+    # The URL to send HTTP POST requests to for MO messages
+    mo_message_url: str
+
+    # Authorization Token to use for the mo_message_url
+    mo_message_url_auth_token: Optional[str] = None
+
+    base_url_path: str = ""
+
+    state_cache_class: str = f"{junebug_state_cache.__name__}.MemoryJunebugStateCache"
+    state_cache_config: dict = field(factory=dict)
+
+    # # Maximum time (seconds) a mo_message_url is allowed to take to process a message
+    # mo_message_url_timeout: int = 10
+    # # Maximum time (seconds) a mo_message_url is allowed to take to process an event
+    # event_url_timeout: int = 10
+
+    def url(self, path: str) -> str:
+        return f"{self.base_url_path}{path}"
```

Comment:
Should we rather use urljoin here instead? Takes care of missing or too many `/`s between the base and the path

Reply:
Good idea.